### PR TITLE
[WIP] Stanford Cars prototype dataset

### PR DESCRIPTION
**Pull request by abhi-glitchhg**
Currently a draft.

### CODE REVIEW

**Comment by abhi-glitchhg** [comment]
I have got a problem and I am not able to fix it. 
the `_make_datapipe` method recieves a list of `IterPipe` objects. There will be two objects becuase `resources` method return a list with two `HttpResource` objects. 

So as per my understanding first object will related to images and second object will be related to labels and bounding boxes. 
And this is correct ; 

`
images dp:  <torchdata.datapipes.iter.util.tararchivereader.TarArchiveReaderIterDataPipe object at 0x7f43f5ed4910> 
 targets dp:  <torch.utils.data.datapipes.iter.fileopener.FileOpenerIterDataPipe object at 0x7f43f5ed4340>`

but sometimes both the dps are related to the labels and bounding boxes;

`images dp:  <torch.utils.data.datapipes.iter.fileopener.FileOpenerIterDataPipe object at 0x7fd7ef58edc0> 
targets dp:  <torch.utils.data.datapipes.iter.fileopener.FileOpenerIterDataPipe object at 0x7fd7ef5a7160>`


I do not understand why this is happening?
Can anyone help me out?
Thanks

cc @pmeier

File: torchvision/prototype/datasets/_builtin/__init__.py [hunk 1]
```diff
@@ -8,10 +8,11 @@
 from .fer2013 import FER2013
 from .gtsrb import GTSRB
 from .imagenet import ImageNet
-from .mnist import MNIST, FashionMNIST, KMNIST, EMNIST, QMNIST
+from .mnist import EMNIST, FashionMNIST, KMNIST, MNIST, QMNIST
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/__init__.py:11]: Could you revert this? I'm guessing you are using `usort>=1`. We are currently bound to `usort==0.6.4` until #5106 is merged.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from torchdata.datapipes.iter import Filter, IterDataPipe, Mapper, Zipper
+from torchvision.prototype.datasets.utils import Dataset, DatasetConfig, DatasetInfo, HttpResource, OnlineResource
+from torchvision.prototype.datasets.utils._internal import hint_sharding, hint_shuffling, path_comparator, read_mat
+from torchvision.prototype.features import BoundingBox, EncodedImage
+
+
+class StanfordCars(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            name="stanford_cars",
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:12]: For now the convention is to use a hyphen to separate words in the dataset name.
```suggestion
            name="stanford-cars",
```

Note that you will also have to rename the `.categories` file and should rename the data folder `~/.cache/torch/datasets/vision/stanford_cars` to avoid re-downloading everything.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        "test_ground_truth": f"{_URL_ROOT}car196/cars_test_annos_withlabels.mat",
+        "devkit": f"{_URL_ROOT}cars/car_devkit.tgz",
+    }
+
+    _CHECKSUM = {
+        "train": "512b227b30e2f0a8aab9e09485786ab4479582073a144998da74d64b801fd288",
+        "test": "bffea656d6f425cba3c91c6d83336e4c5f86c6cffd8975b0f375d3a10da8e243",
+        "test_ground_truth": "790f75be8ea34eeded134cc559332baf23e30e91367e9ddca97d26ed9b895f05",
+        "devkit": "512b227b30e2f0a8aab9e09485786ab4479582073a144998da74d64b801fd288",
+    }
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:33]: I would prefer to have a mapping of the actual file names to the checksums.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+    ) -> IterDataPipe[Dict[str, Any]]:
+
+        images_dp, targets_dp = resource_dps
+        print(config.split)
+        if config.split == "train":
+            targets_dp = Filter(targets_dp, path_comparator("name", "cars_train_annos.mat"))
+        dp = Zipper(images_dp, targets_dp)
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:54]: This is dangerous. Did you make sure that both datapipes are perfectly aligned? Otherwise we should use an `IterKeyZipper`.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        dp = hint_sharding(dp)
+        dp = hint_shuffling(dp)
+        return Mapper(dp, self._read_images_and_labels)
+
+    def _read_images_and_labels(self, data):
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:59]: Nit: this method is canonically called `_prepare_sample` and is placed before the `_make_datapipe` method.

**Review by pmeier — COMMENTED** [review]
Thanks @abhi-glitchhg for reporting this. I've send a patch in #5447 that should fix this behavior. After that is merged, you can update your branch and the issue should be gone.

I had an superficial look and so far this looks really good! I'll do another round after you mark the PR as ready for review.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py (hunk 4 above)
Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py:54]: I wasn't aware of this! I will look into it. Thanks!

**Comment by abhi-glitchhg** [comment]
Sorry for the delay! I will try to finish this by the weekend!

File: torchvision/prototype/datasets/_builtin/stanford_cars.py (hunk 4 above)
Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py]: I think that zipping these two pipelines is not a good idea; Because `targets_dp` contains only one file and images_dp contains around 8000 images;  So with the current implementation, we can access only one image.. 

I was thinking about decoding the `targets_dp` and saving the target_dp data into a variable (something like `self.labels` ) and then accessing it whenever needed...

Could you give me some feedback regarding this? Like should i go with this? Or is there any other way to tackle this..

Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py]: They are aligned :)

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
+from typing import Any, Dict, List, Tuple, Iterator
+
+import numpy as np
+from torchdata.datapipes.iter import Filter, IterDataPipe, Mapper, Zipper
+from torchvision.prototype.datasets.utils import Dataset, DatasetConfig, DatasetInfo, HttpResource
+from torchvision.prototype.datasets.utils._internal import hint_sharding, hint_shuffling, path_comparator, read_mat
+from torchvision.prototype.features import BoundingBox, EncodedImage
+
+
+class _StanfordCarsLabelReader(IterDataPipe[Tuple[np.ndarray, int]]):
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:10]: 1. You can make this "public" since this part of `torchvision` is private as a whole and we only explicitly expose stuff.
2. The generic type annotation has to match the one from `__iter__`. See the comment below for an explanation
3. 
```suggestion
class StanfordCarsLabelReader(IterDataPipe[Tuple[int, int, int, int, int, str]]):
```

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    def __init__(self, datapipe: IterDataPipe[Dict[str, Any]]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[str]]:
+        for _, file in self.datapipe:
+            file = iter(read_mat(file, squeeze_me=True)["annotations"])
+            for line in file:
+                yield line
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:18]: This was hard to parse for me until I run this code myself. Maybe something like this?

```suggestion
    def __iter__(self) -> Iterator[Tuple[int, int, int, int, int, str]]:
        for _, file in self.datapipe:
            data = read_mat(file, squeeze_me=True)
            for ann in data["annotations"]:
                yield tuple(ann)
```

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+    ) -> IterDataPipe[Dict[str, Any]]:
+
+        images_dp, targets_dp = resource_dps
+        if config.split == "train":
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:91]: Can we align this with the check in `self.resources()`? There you check for `config.split == "test"`. No opinion on what check we should use.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+    _CHECKSUM = {
+        "train": "b97deb463af7d58b6bfaa18b2a4de9829f0f79e8ce663dfa9261bf7810e9accd",
+        "test": "bffea656d6f425cba3c91c6d83336e4c5f86c6cffd8975b0f375d3a10da8e243",
+        "cars_test_annos_withlabels": "790f75be8ea34eeded134cc559332baf23e30e91367e9ddca97d26ed9b895f05",
+        "car_devkit": "512b227b30e2f0a8aab9e09485786ab4479582073a144998da74d64b801fd288",
+    }
+
+    def resources(self, config: DatasetConfig) -> List[HttpResource]:
+        resources = [HttpResource(self._URLS[config.split], sha256=self._CHECKSUM[config.split])]
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:48]:
```suggestion
    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
        resources: List[OnlineResource] = [HttpResource(self._URLS[config.split], sha256=self._CHECKSUM[config.split])]
```

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                [
+                    target[0],
+                    target[1],
+                    target[2],
+                    target[3],
+                ],
+                format="xyxy",
+                image_size=image.image_size,
+            ),
+        )
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:81]: In general, this method does what it should do. I've cleaned it up a bit:

1. Not sure why you want to return the index of the image? Is there a concrete use case where this is necessary? If not, the information is available for the user to extract from the path and thus there is no need to return it explicitly.
2. Since the `image_path` is the only path, there is no need to namespace it.
3. We have a special feature type for labels: `Label`. Among the numeric label, it can also store the human readable categories. Plus, canonically, we store the label under the `"label"` key.
4. If the coordinates of the bounding box are already in the right order, you can simply use slice syntax.

```suggestion
    def _prepare_sample(self, data: Tuple[Tuple[str, BinaryIO], Tuple[int, int, int, int, int, str]]) -> Dict[str, Any]:
        image, target = data
        path, buffer = image
        image = EncodedImage.from_file(buffer)

        return dict(
            path=path,
            image=image,
            label=Label(target[4] - 1, categories=self.categories),
            bounding_box=BoundingBox(target[:4], format="xyxy", image_size=image.image_size),
        )
```

File: torchvision/prototype/datasets/_builtin/stanford-cars.categories [hunk 11]
```diff
@@ -0,0 +1,196 @@
+AM General Hummer SUV 2000
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford-cars.categories:1]: How did you generate this file? Unless you need to compile it manually, the best way is to put a `_generate_categories` method on the dataset and use the `torchvision/prototype/datasets/generate_categories_file.py` script. This way we have a record of the code that was used to generate the file.

File: test/builtin_dataset_mocks.py [hunk 12]
```diff
@@ -1390,3 +1390,45 @@ def pcam(info, root, config):
...
+
+
+@register_mock
+def stanford_cars(info, root, config):
+    import scipy.io as io
+    from numpy.core.records import fromarrays
+
+    num_examples = {"train": 5, "test": 7}[config["split"]]
+    num_classes = 3
+    base_folder = pathlib.Path(root)
```
Comment (pmeier) [inline test/builtin_dataset_mocks.py:1402]: `root` is already a `pathlib.Path` so you can remove `base_folder` from this function and use `root` instead.

**Review by pmeier — COMMENTED** [review]
Thanks a lot @abhi-glitchhg! I have some minor comments inline, but otherwise this looks really good!

File: torchvision/prototype/datasets/_builtin/stanford_cars.py (hunk 10 above)
Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py:81]: I was returning the index of the image for testing purposes; I was checking whether the images_dp and target_dp are aligned or not. 
Agree with all the changes that you have suggested!

File: test/builtin_dataset_mocks.py (hunk 12 above)
Comment (abhi-glitchhg) [reply test/builtin_dataset_mocks.py:1402]: Okay! thanks

File: torchvision/prototype/datasets/_builtin/stanford_cars.py (hunk 7 above)
Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py:18]: Oh sorry! My bad! I should have made it more readable.

File: torchvision/prototype/datasets/_builtin/stanford-cars.categories (hunk 11 above)
Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford-cars.categories:1]: I did not add the `generate_categories` method in the dataset because we will have to download `car_devkit.tgz` irrespective of the split(train/test.) 
That's why I did not include it. 
But if you insist, I will :)

Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/stanford-cars.categories:1]: The method is not executed at runtime. So you only need this file locally if you want to generate the file. And this is a compromise I'm willing to make to have a reproducible generation. Make sure you use something like

https://github.com/pytorch/vision/blob/50608fbc263da386ad7abf7b32bd32ed7f691170/torchvision/prototype/datasets/_builtin/imagenet.py#L193-L195

to actually get the resources that include the file you need.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+
+class StanfordCarsLabelReader(IterDataPipe[Tuple[int, int, int, int, int, str]]):
+    def __init__(self, datapipe: IterDataPipe[Dict[str, Any]]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[int, int, int, int, int, str]]:
+        for _, file in self.datapipe:
+            data = read_mat(file, squeeze_me=True)
+            for ann in data["annotations"]:
+                yield tuple(ann)
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:17]: We can just ignore `mypy` here

```suggestion
                yield tuple(ann)  # type: ignore[misc]
```

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        return Mapper(dp, self._prepare_sample)
+
+    def _generate_categories(self, root: pathlib.Path) -> List[str]:
+        config = self.info.make_config(split="train")
+        resources = self.resources(config)
+
+        devkit_dp = resources[1].load(root)
+        meta_dp = Filter(devkit_dp, path_comparator("name", "cars_meta.mat"))
+        print(meta_dp)
+        return Mapper(meta_dp, self.prepare_categories)
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:100]: In this method you don't need to mess around with datapipes unless it is helpful. In this case it makes sense to stay in "datapipe land" until after the filter. Starting from that point you only have a single file that you need to parse.

```suggestion
        _, meta_file = next(iter(meta_dp))

        return list(read_mat(meta_file, squeeze_me=True)["class_names"])
```

> Here i am returning a list of strings which contains the class names

You returned a datapipe that contained a list of strings. This is treated the same way as a list of lists and this is why the generation put everything on one line.

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+            )
+        else:
+            resources.append(HttpResource(url=self._URLS["car_devkit"], sha256=self._CHECKSUM["car_devkit"]))
+
+        return resources
+
+    def prepare_categories(self, dp: IterDataPipe) -> List[str]:
+        _, file = dp
+        classes = list(read_mat(file, squeeze_me=True)["class_names"])
+        return classes
```
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/stanford_cars.py:63]: This is no longer needed.

```suggestion
```

File: torchvision/prototype/datasets/_builtin/stanford_cars.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
```
[the remark this answers is not in the document]
Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py:63]: See https://github.com/pytorch/vision/pull/5442#discussion_r821839019.

[the remark this answers is not in the document]
Comment (abhi-glitchhg) [reply torchvision/prototype/datasets/_builtin/stanford_cars.py:63]: Ohhhhh understood.

forgive me for those mistakes😞

File: test/builtin_dataset_mocks.py [hunk 17]
```diff
@@ -1390,3 +1390,44 @@ def pcam(info, root, config):
...
 
     return num_images
+
+
+@register_mock
+def stanford_cars(info, root, config):
+    import scipy.io as io
+    from numpy.core.records import fromarrays
+
+    num_examples = {"train": 5, "test": 7}[config["split"]]
```
Comment (pmeier) [inline test/builtin_dataset_mocks.py:1400]: Can we call this 

```suggestion
    num_samples = {"train": 5, "test": 7}[config["split"]]
```

to align it with the terminology of the other datasets?

File: test/builtin_dataset_mocks.py [hunk 18]
```diff
@@ -1390,3 +1390,44 @@ def pcam(info, root, config):
...
+    num_classes = 3
```
Comment (pmeier) [inline test/builtin_dataset_mocks.py:1401]: Can we call this

```suggestion
    num_categories = 3
```

to align it with the terminology of the other datasets?

File: test/builtin_dataset_mocks.py [hunk 19]
```diff
@@ -1390,3 +1390,44 @@ def pcam(info, root, config):
...
+    classes = np.random.randint(1, num_classes + 1, num_examples, dtype=np.uint8)
+    fnames = [f"{i:5d}.jpg" for i in range(num_examples)]
+    rec_array = fromarrays(
+        [bbox, bbox, bbox, bbox, classes, fnames],
+        names=["bbox_x1", "bbox_y1", "bbox_x2", "bbox_y2", "class", "fname"],
+    )
+
+    io.savemat(annotations_mat_path, {"annotations": rec_array})
+    if config.split == "train":
+        make_tar(root, "car_devkit.tgz", devkit)
```
Comment (pmeier) [inline test/builtin_dataset_mocks.py:1431]:
```suggestion
        make_tar(root, "car_devkit.tgz", devkit, compression="gz")
```

**Review by pmeier — COMMENTED** [review]
Thanks @abhi-glitchhg, we are almost done! Just a few minor comments including one from a previous review left. After that we are good to merge!

File: test/builtin_dataset_mocks.py [hunk 20]
```diff
@@ -370,11 +370,11 @@ def create_ann_file(root, name):
 
         scipy.io.savemat(str(pathlib.Path(root) / name), dict(box_coord=box_coord, obj_contour=obj_contour))
 
-    def create_ann_folder(root, name, file_name_fn, num_examples):
```
Comment (abhi-glitchhg) [inline test/builtin_dataset_mocks.py:373]: Ooops!

**Review by pmeier — APPROVED** [review]
Awesome, thanks a lot for the patience @abhi-glitchhg!

**Comment by abhi-glitchhg** [comment]
@pmeier  thanks for the guidance! <3

**Comment by pmeier** [comment]
Hey @abhi-glitchhg, we appreciate your responsiveness, but you don't need to update the branch for every new commit on the `main` branch. We will do that before we merge. Otherwise, we would be wasting a lot of CI resources for very little gain. 

One exception is if GitHub auto-detects some conflicts. In that case, feel free to address them before the final approval.

**Review by NicolasHug — APPROVED** [review]
Stamping, thanks @abhi-glitchhg
